### PR TITLE
Add environment variable to control host memory allocation in `hostdevice_vector`

### PR DESCRIPTION
**Pull request by vuule**
## Description
Enables experimentation related to issue https://github.com/rapidsai/cudf/issues/12186

Frequent allocation of pinned memory becomes costly in multi-GPU environment, where all devices synchronize and allocation creates huge bubbles in GPU utilization.

This PR adds environment variable `LIBCUDF_PREFER_PAGEABLE_MEMORY` that controls whether `hostdevice_vector` uses pinned memory for the host side. `hostdevice_vector` makes up most of pinned memory allocations in cuIO, so this environment variable should greatly impact the execution when using multiple GPUs.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 1]
```diff
@@ -51,27 +55,41 @@ class hostdevice_vector {
...
+      h_data_owner = thrust::host_vector<T>();
+      std::get<0>(h_data_owner).reserve(max_size);
+      std::get<0>(h_data_owner).resize(initial_size);
+      host_data = std::get<0>(h_data_owner).data();
+    } else {
+      h_data_owner = thrust::host_vector<T, cudf::detail::pinned_allocator<T>>();
+      std::get<1>(h_data_owner).reserve(max_size);
+      std::get<1>(h_data_owner).resize(initial_size);
+      host_data = std::get<1>(h_data_owner).data();
+    }
```
Comment (ttnghia) [inline cpp/src/io/utilities/hostdevice_vector.hpp:71]:
```suggestion
    auto const allocate_data = [&] (auto& h_data) {
      h_data.reserve(max_size);
      h_data.resize(initial_size);
      host_data = h_data.data();
    };

    if (use_pageable_buffer) {
      h_data_owner = thrust::host_vector<T>();
      allocate_data(std::get<0>(h_data_owner));
    } else {
      h_data_owner = thrust::host_vector<T, cudf::detail::pinned_allocator<T>>();
      allocate_data(std::get<1>(h_data_owner));
    }
```

**Review by ttnghia — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Review by hyperbolic2346 — APPROVED** [review]

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 2]
```diff
@@ -51,27 +55,43 @@ class hostdevice_vector {
     : d_data(0, stream)
   {
     CUDF_EXPECTS(initial_size <= max_size, "initial_size cannot be larger than max_size");
-    h_data.reserve(max_size);
-    h_data.resize(initial_size);
+
+    auto const use_pageable_buffer =
+      cudf::io::detail::getenv_or("LIBCUDF_PREFER_PAGEABLE_MEMORY", 0);
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/utilities/hostdevice_vector.hpp:60]: How about `LIBCUDF_IO_PREFER_PAGEABLE_TMP_MEMORY`?

**Review by robertmaynard — APPROVED** [review]

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**